### PR TITLE
[dogstatsd] ignore origin tags if entityID is present

### PR DESCRIPTION
**Pull request by clamoriniere**
### What does this PR do?

refactor the originTags detection in dogstatsd server:
* ignore originTags if the entityID tags is present
* ignore entityID value equal to "none"

### Motivation

avoid having the origin tags if the entity ID tag is present in the metrics.

### Additional Notes

this PR is linked to: https://github.com/DataDog/jmxfetch/pull/264

### CODE REVIEW

**Review by vboulineau — DISMISSED** [review]

**Review by olivielpeau — COMMENTED** [review]
LGTM overall, thanks for doing this!

A few comments:
* this PR is making changes to the hot loop of dogstatsd, so let's make sure we're not introducing perf regressions: instead of finding the origin tags once per packet, we now find them once per message, so potentially a lot more (by a factor of 10 to 1000). Do you know if this has an impact on performance? Alternatively we could find the tags once and pass them down the function calls.
* could you add unit tests on the new behavior of `enrichTags`?
* could you add some details in the PR about the reasons why we want these 2 new behaviors? (typically I fully understand the `dd.internal.entity_id:none` case, but I'm not sure I understand the other one, and in any case I think it makes sense to describe both)

**Comment by clamoriniere** [comment]
@olivielpeau PR updated with better optimization and tests

File: pkg/dogstatsd/server.go
Comment (olivielpeau) [inline pkg/dogstatsd/server.go]:
```suggestion
	// we don't use "sync.Once" here because we know only one goroutine can call the function `getTags()`
```

File: pkg/dogstatsd/enrich.go
Comment (olivielpeau) [inline pkg/dogstatsd/enrich.go]: let's move this back to `server.go`? (since it's only used there)

**Review by olivielpeau — DISMISSED** [review]
A couple of nits inline, but LGTM overall, feel free to merge once nits are addressed.

Before merging:
* could you check this PR has no performance impact on this benchmark? https://github.com/DataDog/datadog-agent/blob/ae69351d3ed0eccc33fd3672a620a628d55375e3/pkg/dogstatsd/server_bench_test.go#L49-L74 (especially in terms of memory allocations)
* could you describe why this behavior: `ignore originTags if the entityID tags is present` makes sense? (the jmxfetch PR always adds a value of `none` so I'm not sure I understand this other case)

**Comment by clamoriniere** [comment]
> * could you describe why this behavior: `ignore originTags if the entityID tags is present` makes sense? (the jmxfetch PR always adds a value of `none` so I'm not sure I understand this other case)

JMXFetch only send "none" if the communication with dogstatsd is done with the socket. in this case we know that will get entity tag, but we want to avoid it.

**Comment by clamoriniere** [comment]
benchmark result

**Review by olivielpeau — APPROVED** [review]
not sure where this additional alloc comes from, but let's follow up on that later.

LGTM ✅

**Pull request merged**
